### ISSUE TITLE
2.4.0 may refuse to build if the indirect dependencies have not main/module/exports specified in its package.json

### ISSUE THREAD

**Issue by thislight** [solidjs/vite-plugin-solid#70]
Reproducing steps:
````
mkdir my-solid-app && cd my-solid-app
pnpm create solid # without SSR, use TypeScript
pnpm add solid-hcaptcha
pnpm dev
````
- Create a new app
- Add a dependency which have a type declaration package as a direct dependency (not dev dependency)
- Run dev server
- Will saw an error like
````
Error: Failed to resolve entry for package "@hcaptcha/types". The package may have incorrect main/module/exports specified in its package.json.
````
[(`@hcaptcha/types` is a type declaration package)](https://www.npmjs.com/package/@hcaptcha/types)

It should be built without problem and 2.3.11 did. Override the version to 2.3.11 in package.json and do `pnpm i`:
````json
{
  "pnpm": {
    "overrides": {
      "vite-plugin-solid": "2.3.11"
    }
  }
}
````
And it works.

Environment:
````
$ node -v
v18.1.0
$ pnpm -v
v7.14.0
````

**Comment by thislight** [comment]
I dig further and make a smaller reproduce:
````json
{
    "name": "another_package",
    "version": "0.1.0",
    "index": "index.js",
    "dependencies": {
        "@hcaptcha/types": "^1.0.3"
    },
    "peerDependencies": {
        "solid-js": "^1.6.0"
    },
    "exports": {
        ".": {
            "solid": "./index.js"
        }
    }
}
````
The app will be break if it depends on this `another_package`.

**Comment by thislight** [comment]
I might figure out what's wrong here.
2.4.0 includes a breaking behaviour from vitefu, see:

https://github.com/solidjs/vite-plugin-solid/blob/bf5dd1ad3d9588359952b0627928558b35240329/src/index.ts#L304

In 2.3.11, the only `optimizeDeps.include` is from nestedDeps, now it also comes from the result of vitefu's `crawlFrameworkPkgs`. The problem comes from:

https://github.com/svitejs/vitefu/blob/d96945bd29858db7b1d6f915dad1a91ebe7fb1fc/src/index.js#L169

It will take a chance to be added to `optimizeDepsInclude` if the package is a dependencies of a "framework package", in the case, the "framework packages" are the packages contains field `exports.solid` in their package.json.

The first question is, when? Let's take look at the `crawl` function, there are only one situation to be scan as dependencies：
- The package is a direct dependency of a "framework package".
    https://github.com/svitejs/vitefu/blob/d96945bd29858db7b1d6f915dad1a91ebe7fb1fc/src/index.js#L146
    https://github.com/svitejs/vitefu/blob/d96945bd29858db7b1d6f915dad1a91ebe7fb1fc/src/index.js#L160

And the `pkgJsonNeedsOptimization` is defined as:
````js
export function pkgJsonNeedsOptimization(pkgJson) {
  // only optimize if is cjs, using the below as heuristic
  // see https://github.com/sveltejs/vite-plugin-svelte/issues/162
  if (pkgJson.module || pkgJson.exports) return false
  // has implicit index.js entrypoint, prebundle
  // see https://github.com/sveltejs/vite-plugin-svelte/issues/281
  if (!pkgJson.main) return true // (1)
  // ensure entry is js so vite can prebundle it
  // see https://github.com/sveltejs/vite-plugin-svelte/issues/233
  const entryExt = path.extname(pkgJson.main)
  return !entryExt || entryExt === '.js' || entryExt === '.cjs'
}
````
Look at the (1), this function assume the package have a entry point even if they have not a explicit one. In actual cases, no explicit entry point doesn't mean it does have one implicitly.

Well, we found the cause is the unexpected include in `optimizeDeps.include`. So we just need a explicit exclusion for our problem, right?
Well, that is another problem.
vitefu's `crawlFrameworkPkgs` does have a section try to respect user configuration of `optimizeDeps.exclude`:

https://github.com/svitejs/vitefu/blob/d96945bd29858db7b1d6f915dad1a91ebe7fb1fc/src/index.js#L34

But it does not work automatically, it reads from `options.viteUserConfig`, a function argument. Unfortunately...`vite-solid-plugin` does not feed this field:
https://github.com/solidjs/vite-plugin-solid/blob/bf5dd1ad3d9588359952b0627928558b35240329/src/index.ts#L279

So this bug will prevent user use `optimizeDeps.exclude` to exclude problems.

Conclusion:
- `vite-plugin-solid` should feed user configuration to vitefu `crawlFrameworkPkgs`
- vitefu should check if the package actually have one implicit entry point before adding it to `optimizeDeps.include`
- type declaration package should not be added as (not dev) dependency (for solid-hcaptcha)

**Comment by bluwy** [comment]
Released vitefu `v0.2.0` with the fixes. Thanks!
